GH-39811: [R] better documentation for col_types argument in open_delim_dataset

atsyplenkov opened the pull request:
### Rationale for this change
Hi, can you please consider this tiny update to the docs? In the current documentation, it's misleading how to specify col_types when a delimited file is scanned using `open_csv_dataset`, `open_delim_dataset`, etc. Reading what is currently written, one may assume that they can declare column types by providing the compact string representation that `readr` uses. 

https://github.com/apache/arrow/blob/3c8fe098c7f5e0e40bd06bc6afca8412eb81f56e/r/man/open_delim_dataset.Rd#L164-L165

But it doesn't work. See reprex below

```r
library(arrow)
#> 
#> Attaching package: 'arrow'
#> The following object is masked from 'package:utils':
#> 
#>     timestamp
tf <- tempfile()
dir.create(tf)
df <- data.frame(x = c("1", "2", "NULL"))

file_path <- file.path(tf, "file1.txt")
write.table(df, file_path, sep = ",", row.names = FALSE)

open_csv_dataset(file_path, na = c("", "NA", "NULL"), col_types = "c")
#> Error:
#> ! Unsupported `col_types` specification.
#> ℹ `col_types` must be NULL, or a <Schema>.

unlink(tf)
```

### What changes are included in this PR?
The current PR provides a clearer explanation of what should be passed to the `col_types` argument, along with a basic example for the `open_csv_dataset()`.

### Are these changes tested?
Not needed, as only the R documentation has been updated

### Are there any user-facing changes?
Only the R documentation has been updated

atsyplenkov commented:
@jonkeane sure, I'll see what can I do there

atsyplenkov commented:
@jonkeane sorry for the long turnaround; this issue fell through the cracks. I followed the approach suggested by @thisisnic. I added a number of tests, and as far as I can see, everything is working.

Tests actually uncovered some problems with handling NA values, I will raise an issue regarding that soon

thisisnic reviewed (changes requested):
Hi @atsyplenkov, thanks for making this PR to fix this, and I appreciate all the work you've put in here.  Getting this merged will improve usability for lots of people, so this is great!

The main changes I've suggested here are mostly around moving copied code into helper functions instead.  I've added a few comments around the tests, but also I think we can cut quite a few of them out - I'll give the tests a proper review once we've got the final shape of the changes to the code.

atsyplenkov commented:
@thisisnic thanks for reviewing the code! 

I have made the following changes:
* added a new helper function in util.R, called `parse_compact_col_spec()`, and tests for it in `test-util.R`
* refactored `readr_to_csv_convert_options()` and `check_csv_file_format_args()` to handle this new helper
* updated the tests as requested

thisisnic commented:
Thanks for your changes @atsyplenkov, this is really great!  I refactored a few bits to simplify things a little bit, and I'll wrap up the rest of this PR and merge it once I've ironed out a few kinks.  Cheers!
